modules/imgcodecs/src/grfmt_tiff.cpp

```diff
@@ -480,25 +481,67 @@ bool  TiffDecoder::readData( Mat& img )
             const uint64_t MAX_TILE_SIZE = (CV_BIG_UINT(1) << 30);
             CV_CheckLE((int)ncn, 4, "");
             CV_CheckLE((int)bpp, 64, "");
-            CV_Assert(((uint64_t)tile_width0 * tile_height0 * ncn * std::max(1, (int)(bpp / bitsPerByte)) < MAX_TILE_SIZE) && "TIFF tile size is too large: >= 1Gb");
 
+            bool doReadScanline = false;
             if (dst_bpp == 8)
             {
                 // we will use TIFFReadRGBA* functions, so allocate temporary buffer for 32bit RGBA
                 bpp = 8;
                 ncn = 4;
             }
+            else if (dst_bpp == 16)
+            {
+                // if buffer_size >= MAX_TILE_SIZE, try to use TIFFReadScanline function.
+                if (
+                    (uint64_t)tile_width0 * tile_height0 * ncn * std::max(1, (int)(bpp / bitsPerByte))
+                    >=
+                    MAX_TILE_SIZE
+                )
+                {
+                    uint16_t planerConfig = (uint16)-1;
+                    CV_TIFF_CHECK_CALL(TIFFGetField(tif, TIFFTAG_PLANARCONFIG, &planerConfig));
+
+                    doReadScanline = (!is_tiled)
+                                     &&
+                                     (
+                                         (photometric == PHOTOMETRIC_MINISWHITE)
+                                         ||
+                                         (photometric == PHOTOMETRIC_MINISBLACK)
+                                         ||
+                                         (photometric == PHOTOMETRIC_RGB)
+                                     )
+                                     &&
+                                     (planerConfig != PLANARCONFIG_SEPARATE);
+
+                    if ( doReadScanline )
+                    {
+                        // Read each scanlines.
+                        tile_height0 = 1;
+                    }
+                }
+            }
             else if (dst_bpp == 32 || dst_bpp == 64)
             {
                 CV_Assert(ncn == img.channels());
                 CV_TIFF_CHECK_CALL(TIFFSetField(tif, TIFFTAG_SAMPLEFORMAT, SAMPLEFORMAT_IEEEFP));
             }
+
+            CV_Assert(((uint64_t)tile_width0 * tile_height0 * ncn * std::max(1, (int)(bpp / bitsPerByte)) < MAX_TILE_SIZE) && "TIFF tile size is too large: >= 1Gb");
```

Comment:
`win32` build emits this - is it expected?
(however I'm not sure about stdout/stderr mess on windows system)

```
[ RUN      ] Imgcodecs_Tiff.decode_tile_remainder
imread_('C:\build\precommit_windows32\build\_ocl_tmp\__opencv_temp.s92luevb\ocvF424.tmp.tiff'): can't read data: OpenCV(3.4.18-dev) C:\build\precommit_windows32\3.4\opencv\modules\imgcodecs\src\grfmt_tiff.cpp:529: error: (-215:Assertion failed) ((uint64_t)tile_width0 * tile_height0 * ncn * std::max(1, (int)(bpp / bitsPerByte)) < MAX_TILE_SIZE) && "TIFF tile size is too large: >= 1Gb" in function 'cv::TiffDecoder::readData'

[       OK ] Imgcodecs_Tiff.decode_tile_remainder (85 ms)
```

http://pullrequest.opencv.org/buildbot/builders/precommit_windows32/builds/100061/steps/test_imgcodecs/logs/stdio

Reply:
Linux32 logs: 

```
[ RUN      ] Imgcodecs_Tiff.decode_tile16384x16384
imread_('/tmp/__opencv_temp.hjunrS/__opencv_temp.DTpYok.tiff'): can't read data: OpenCV(3.4.18-dev) /build/precommit_linux32/3.4/opencv/modules/imgcodecs/src/grfmt_tiff.cpp:529: error: (-215:Assertion failed) ((uint64_t)tile_width0 * tile_height0 * ncn * std::max(1, (int)(bpp / bitsPerByte)) < MAX_TILE_SIZE) && "TIFF tile size is too large: >= 1Gb" in function 'readData'

[       OK ] Imgcodecs_Tiff.decode_tile16384x16384 (11528 ms)
```

http://pullrequest.opencv.org/buildbot/builders/precommit_linux32/builds/100035/steps/test_imgcodecs/logs/stdio

Reply:
## decode_tile_reminder

 `decode_tile_reminder` test is a test to decode tiff file in OpenCV extra. I see source code and I think temporary files are not used in tests. But error log shows used it.

https://github.com/opencv/opencv/blob/3.4.18/modules/imgcodecs/test/test_tiff.cpp#L96

I feel it's a strange phenomenon. I have not been able to reproduce it on linux32. To investigate it, I believe more testing with the new patch is necessary.

Reply:
## decode_tile16384x16384

I believe the `decode_tile16384x16384` test is affected by code that fixes a potential problem.

Currently, (a) the buffer size that is about to be allocated and (b) the buffer size that is actually allocated are different. Therefore, it can cause out of memory.

I think the new patch has fixed it.

### 3.4.18
https://github.com/opencv/opencv/blob/3.4.18/modules/imgcodecs/src/grfmt_tiff.cpp#L498

- Step 1. (Line 483) Verify that the memory size to be allocated by buffer does not exceed 1GB.
- Step 2. (Line 488-489) If dst_bpp=8, update to bpp=8, ncn=4 (to use TIFFReadRGBA* functions).
- Step 3. (Line 496) Calculate buffer_size.
- Step 4. (Line 497) Allocate buffer.

For example, consider executing imread( CV_8UC1 image file, IMREAD_COLOR).

- Step1: Size check is done in 8bit GRAY color space.
- Step2: bpp/ncn is rewritten.
- Step3: Calculated buffer_size is based on 24bit RGBA color space.
- Result: The actual amount of memory to allocate exceeds 1GB.

### How to fix

I think it should be modified as follows.

- Step1: If dst_bpp=8, update to bpp=8, ncn=4 (to use TIFFReadRGBA* functions).
- Step2: Calculate buffer_size.
- Step 3: Verify if buffer_size does not exceed 1GB.
- Step 4: allocate buffer.